# 第一次作业

## 1.数值误差的避免

### (a).
设 $\epsilon=\frac{1}{2}\epsilon_m$，这里将$fl(x)=x(1+\delta)$中的$\delta$取为最大值$\epsilon$

则$\bigoplus_{i=1}^N x_i$

$=(((x_1+x_2)(1+\epsilon)+x_3)(1+\epsilon)+\cdots+x_N)(1+\epsilon)$

$=(x_1+x_2)(1+\epsilon)^{n-1}+x_3(1+\epsilon)^{N-2}+\cdots+x_N(1+\epsilon)$

从而$\frac{\sum_{i=1}^N x_i}{N}$的舍入结果为$\frac{(x_1+x_2)(1+\epsilon)^{n-1}+x_3(1+\epsilon)^{N-2}+\cdots+x_N(1+\epsilon)}{N}(1+\epsilon)$

于是$\bar{x}$的舍入误差为$\frac{(x_1+x_2)(1+\epsilon)^{n-1}+x_3(1+\epsilon)^{N-2}+\cdots+x_N(1+\epsilon)}{N}(1+\epsilon)-\frac{\sum_{i=1}^N x_i}{N}$

$=\frac{(x_1+x_2)n\epsilon+x_3(n-1)\epsilon+\cdots+x_N\cdot2\epsilon}{N}$

$\le x_{max}\frac{2n\epsilon+(n-1)\epsilon+(n-2)\epsilon+\cdots+2\epsilon}{N}$

$=\frac{x_{max}}{2}(n+3-\frac{2}{n})\epsilon$

$\simeq \frac{x_{max}(n+3)}{2}\cdot \frac{\epsilon_M}{2}$



### (b).

第二个。因为第一个式子涉及两个大数的相减,这会导致问题的条件数急剧变大，算法变得不稳定

事实上记二者结果为$s_1,s_2$，$|\frac{\partial s_1}{\partial x_i}|=|\frac{2x_i}{N-1}|\ge|\frac{\partial s_2}{\partial x_i}|=|\frac{2(x_i-\bar{x})}{N-1}|$,

$|\frac{\partial s_1}{\bar{x}}|=|\frac{-2N\bar{x}}{N-1}|\ge|\frac{\partial s_2}{\bar{x}}|=|\frac{1}{N-1}\sum_{i=1}^{N}2(x_i-\bar{x})|=0$

从而由$cond=|\frac{(f(\hat{x})-f(x))/f(x)}{(\hat{x}-x)/x}|$知第二种条件数小，算法稳定。

### (c).

证明：

$I_0=\int_0^1 \frac{1}{x+5} \, dx=\ln \left(\frac{6}{5}\right)$,

$I_k+5 I_{k-1}=\int_0^1 \frac{5 x^{k-1}+x^k}{x+5} \, dx=\int_0^1 x^{k-1} \, dx=\frac{1}{k}$.

若计算$I_0$时有一微小误差$\epsilon_0=\epsilon$，设由此导致的$I_k$的误差为$\epsilon_k$，有

$I_k+5 I_{k-1}=\frac{1}{k}$, &emsp;&emsp; $(I_k+\epsilon_k)+5 (I_{k-1}+\epsilon_{k-1})=\frac{1}{k}$


两式相减即得
$\epsilon_k=-5\epsilon_{k-1}$

于是
$|\epsilon_n|=5^n |\epsilon|$

由此可见，随着n的增加，因$\epsilon$导致的误差被无限放大，故由上述算法计算得到的$I_n$是不稳定的


## 2. 矩阵的模与条件数

### (a).

我们可以按行列式的最后一行展开，有

$A=
\left|
\begin{matrix}
 1       & -1      &-1    & \cdots  & -1      \\
 0       & 1       &-1   & \cdots & -1      \\
 \vdots    &\vdots    &\vdots & \ddots & \vdots \\
 0       & 0       & 0   & \cdots & 1      \\
\end{matrix}
\right|_{n\times n}
=
\left|
\begin{matrix}
 1       & -1      &-1    & \cdots  & -1      \\
 0       & 1       &-1   & \cdots & -1      \\
 \vdots    &\vdots    &\vdots & \ddots & \vdots \\
 0       & 0       & 0   & \cdots & 1      \\
\end{matrix}
\right|_{(n-1)\times (n-1)}
=\cdots =
\left|
\begin{matrix}
 1       & -1      \\
 0       & 1      \\
\end{matrix}
\right|
=1$

故A 的确不是奇异矩阵。

### (b).
$n\ge 3$时，有
$A^{-1}=
\left[
\begin{matrix}
 1       & 1      &2     &4    &8    &\cdots  & 2^{n-2}      \\
 0       & 1      &1     &2    &4    &\cdots   & 2^{n-3}    \\
 0       &0      & 1      &1     &2    &\cdots   & 2^{n-4} \\
 \vdots    &\vdots  &\vdots  &\vdots  &\vdots &\ddots & \vdots \\
 0       & 0    & 0   & 0   & 0  & \cdots & 1      \\
\end{matrix}
\right]_{n\times n}$


$n=1$时，
$A^{-1}=
\left[
\begin{matrix}
 1 
\end{matrix}
\right]$，
$n=2$时，
$A^{-1}=
\left[
\begin{matrix}
 1 &1 \\ 0 &1
\end{matrix}
\right]$.

### (c).

证明：

因为 $||\overrightarrow{x}||_\infty=max\{|x_i|\}$

所以$||A||_\infty=\frac{max\{\sum_{j=1}^n a_{ij}x_j\}}{max\{x_j\}}=max\{\sum_{j=1}^na_{ij}\frac{x_j}{max\{x_j\}}\}\le
max\{\sum_{j=1}^n|a_{ij}|\frac{|x_j|}{max\{|x_j|\}}\}\le max\{\sum_{j=1}^n|a_{ij}|\}$

等号在$|x_1|=|x_2|=\cdots=|x_n|$，$x_ja_{ij}>0,j=1,2,\cdots n$时成立，故$||A||_\infty=max\{\sum_{j=1}^n|a_{ij}|\}$

### (d).

对于p=2的模，$||B||_2=\sqrt{\rho(B^\dagger B)}$，其中对厄米矩阵A，$\rho{A}=max\{|\lambda_i|\}$，$\lambda_i$为A的特征值。

由于幺正矩阵$U^\dagger U=UU^\dagger=I$，故$||U||_2=||U^\dagger||_2=1$

$||UA||_2=\sqrt{\rho((UA)^\dagger (UA))}=\sqrt{\rho(A^\dagger U^\dagger UA)}=\sqrt{\rho(A^\dagger A)}=||A||_2$

从而$K_2(UA)=||UA||_2||(UA)^-1||_2=||A||_2||A^-1||_2=K_2(A)$

### (e).
对于上述矩阵，$||A||_\infty=1+1+1+\cdots+1=n$，$||A^-1||_\infty=1+1+2+4+\cdots+2^{n-2}=2^{n-1}$

$K_\infty(A)=n2^{n-1}$

## 3.Hilbert 矩阵

### (a).
由$D=\int_{0}^{1}dx(\sum_{i=1}^{n}c_{i}x^{i-1}-f(x))^2$，为使$D$取最小值，我们要$\frac{\partial D}{\partial c_{j}}=0$

即$\int_0^1dx\cdot 2(\sum_{i=1}^{n}c_{i}x^{i-1}-f(x))x^{j-1}=0$，$\sum_{i=1}^n c_i \int_0^1 x^{i+j-2}dx=\int_0^1x^{j-1}f(x)dx$

即$\sum_{i=1}^n \frac{1}{i+j-1}c_i=\int_0^1x^{j-1}f(x)dx$

所以$(H_n)_{ij}=\frac{1}{i+j-1}$，$b_i=\int_0^1x^{i-1}f(x)dx$

### (b).
${}c^tAc=\sum_{1\leq i,j\leq n}\frac{c_ic_j}{i+j-1}=\sum_{1\leq i,j\leq n}c_ic_j\int_0^1t^{i+j-2}dt=\int_0^1\left(\sum_{i=1}^n c_it^{i-1}\right)^2dt\ge0$

等号成立时有$c_i=0$，$i=1,2,3,\cdots$

对于一个实正定矩阵，其特征根都为正实数，故其行列式必定大于零，从而希尔伯特矩阵非奇异。

### (c).

由斯特林公式，$n!\simeq \sqrt{2 \pi n}(\frac{n}{e})^n$，两边取对数有$\ln(n!) \simeq \frac{1}{2}\ln(2\pi)+(n+\frac{1}{2})\ln(n)-n$，
于是得以快速估计$det(H_n)$近似值如下：

In [1]:
import numpy as np
import pandas as pd

# 计算ln(n!)
def log_frac(n): return np.log(2 * np.pi) / 2 + (n + 1 / 2) * np.log(n) - n

# 计算ln(c_n)
def log_c(n): return sum([log_frac(i) for i in range(1, n)])

# 计算det(H_n)
def det_H(n): return np.exp(4 * log_c(n) - log_c(2 * n))

h_n=[det_H(i) for i in range(2,11)]
pd.DataFrame(columns=(["n"]+[i for i in range(2,11)]),data=[["H(n)"]+h_n])


n         2         3             4             5             6  \
0  H(n)  0.070013  0.000342  1.122800e-07  2.389394e-12  3.251519e-18   

              7             8             9            10  
0  2.808616e-25  1.533334e-33  5.276367e-43  1.142321e-53

### (d).
算法代码如下：

In [2]:
# 上三角矩阵的反代求解
def Solve1(u, b):
    n = u.shape[0]
    x = np.zeros(n, dtype=float)
    x[n - 1] = b[n - 1] / u[n - 1, n - 1]
    for i in range(n - 2, -1, -1):
        x[i] = b[i]
        for j in range(i + 1, n):
            x[i] -= u[i, j] * x[j]
        x[i] /= u[i, i]
    return x


# 下三角矩阵的反代求解
def Solve2(u, b):
    n = u.shape[0]
    x = np.zeros(n, dtype=float)
    x[0] = b[0] / u[0, 0]
    for i in range(1, n):
        x[i] = b[i]
        for j in range(0, i):
            x[i] -= u[i, j] * x[j]
        x[i] /= u[i, i]
    return x


# Gauss消元法，将a变成上三角矩阵，
def GEM(aa, bb): 
    
    # 使用copy避免改变了原有数据
    a = np.copy(aa)
    b = np.copy(bb)
    n = a.shape[0]
    
    for i in range(0, n):
        
        # 支点遴选
        m = np.abs(a[i, i])
        id = i
        for j in range(i + 1, n):
            if np.abs(a[j, i]) > m: # 找最大的|a_ji|
                m = np.abs(a[i, j])
                id = j 
        if i != id: # 将两行互换
            a[[i, id], :] = a[[id, i], :]
            b[[i, id]] = b[[id, i]]
        
        # 进行消元
        for j in range(i + 1, n): # a_jk
            for k in range(i + 1, n): 
                a[j, k] -= a[i, k] * a[j, i] / a[i, i]
            b[j] -= b[i] * a[j, i] / a[i, i]
            a[j, i] = 0
    return (a, b)


# Gauss消元法解方程
def Solve_GEM(a, b):
    u, b = GEM(a, b) # 进行高斯消元
    return Solve1(u, b) # 反解下三角矩阵


# Cholesky分解，得到的h为下三角矩阵，其转置为上三角矩阵
def Cholesky(a):
    
    n = a.shape[0]
    h = np.zeros([n, n])
    h[0, 0] = np.sqrt(a[0, 0]) # h_11
    
    for i in range(1, n):
        
        for j in range(0, i): # h_ij
            h[i, j] = a[i, j]
            for k in range(0, j):
                h[i, j] -= h[i, k] * h[j, k]
            h[i, j] /= h[j, j]
            
        h[i, i] = a[i, i] 
        for k in range(0, i): # h_ii
            h[i, i] -= h[i, k] ** 2
        h[i, i] = np.sqrt(h[i, i])
    return h


# Cholesky分解解方程
def Solve_Cholesky(a, b):
    h = Cholesky(a) # 对a进行Cholesky分解
    return Solve1(np.transpose(h), Solve2(h, b)) # 分别反解一个上三角矩阵和下三角矩阵


# Hilbert矩阵的生成
def Hilbert(n):
    h = np.zeros([n, n], dtype=float)
    for i in range(0, n):
        for j in range(0, n):
            h[i, j] = 1 / (i + j + 1)
    return h

# b_n的生成
def b(n): return np.array([1. for i in range(n)])


计算结果如下：

(1)使用Gauss消元法的结果：

In [3]:
colu=["阶数","x1","x2","x3","x4","x5","x6","x7","x8","x9","x10"]
result1=[[i]+np.ndarray.tolist(Solve_GEM(Hilbert(i),b(i))) for i in range(2,11)]
pd.DataFrame(columns=colu,data=result1)


阶数        x1          x2            x3             x4            x5  \
0   2 -2.000000    6.000000           NaN            NaN           NaN   
1   3  3.000000  -24.000000     30.000000            NaN           NaN   
2   4 -4.000000   60.000000   -180.000000     140.000000           NaN   
3   5  5.000000 -120.000000    630.000000   -1120.000000  6.300000e+02   
4   6 -6.000000  210.000000  -1680.000000    5040.000001 -6.300000e+03   
5   7  7.000000 -336.000002   3780.000018  -16800.000069  3.465000e+04   
6   8 -8.000000  503.999977  -7559.999668   46199.998087 -1.386000e+05   
7   9  8.999925 -719.994729  13859.909635 -110879.348480  4.504476e+05   
8  10 -9.997088  989.747831 -23754.619465  240191.019837 -1.261026e+06   

             x6            x7            x8            x9            x10  
0           NaN           NaN           NaN           NaN            NaN  
1           NaN           NaN           NaN           NaN            NaN  
2           NaN           NaN           NaN           NaN            NaN  
3           NaN           NaN           NaN           NaN            NaN  
4  2.772000e+03           NaN           NaN           NaN            NaN  
5 -3.326400e+04  1.201200e+04           NaN           NaN            NaN  
6  2.162160e+05 -1.681680e+05  5.148000e+04           NaN            NaN  
7 -1.009003e+06  1.261254e+06 -8.236765e+05  2.187891e+05            NaN  
8  3.783137e+06 -6.725664e+06  7.000259e+06 -3.937683e+06  923661.969484

(2)使用Cholesky分解的结果如下：

In [4]:
result2=[[i]+np.ndarray.tolist(Solve_Cholesky(Hilbert(i),b(i))) for i in range(2,11)]
pd.DataFrame(columns=colu,data=result2)


阶数        x1          x2            x3             x4            x5  \
0   2 -2.000000    6.000000           NaN            NaN           NaN   
1   3  3.000000  -24.000000     30.000000            NaN           NaN   
2   4 -4.000000   60.000000   -180.000000     140.000000           NaN   
3   5  5.000000 -120.000000    630.000000   -1120.000000  6.300000e+02   
4   6 -6.000000  210.000000  -1680.000000    5040.000001 -6.300000e+03   
5   7  7.000000 -336.000002   3780.000017  -16800.000064  3.465000e+04   
6   8 -8.000000  503.999988  -7559.999823   46199.998944 -1.386000e+05   
7   9  8.999939 -719.995739  13859.926897 -110879.472597  4.504480e+05   
8  10 -9.997668  989.798372 -23755.703409  240200.928827 -1.261074e+06   

             x6            x7            x8            x9            x10  
0           NaN           NaN           NaN           NaN            NaN  
1           NaN           NaN           NaN           NaN            NaN  
2           NaN           NaN           NaN           NaN            NaN  
3           NaN           NaN           NaN           NaN            NaN  
4  2.772000e+03           NaN           NaN           NaN            NaN  
5 -3.326400e+04  1.201200e+04           NaN           NaN            NaN  
6  2.162160e+05 -1.681680e+05  5.148000e+04           NaN            NaN  
7 -1.009004e+06  1.261255e+06 -8.236772e+05  2.187893e+05            NaN  
8  3.783268e+06 -6.725879e+06  7.000468e+06 -3.937793e+06  923686.208917

实际上，这些方程组的解是整数的，调用Mathematica我们可以得到精确解如下：

(以下代码需要Mathematica 11.0+的环境才能运行)

In [5]:
import subprocess
code='StringReplace[ToString@Table[{i}~Join~LinearSolve[HilbertMatrix[i], Array[1 &, i]], {i, 2, 10}], {\"{\" ->\"[\", \"}" -> \"]\"}]'
st=subprocess.check_output(["wolframscript", "-code",code])
pd.DataFrame(columns=colu,data=eval(st))


阶数  x1   x2       x3        x4         x5         x6         x7         x8  \
0   2  -2    6      NaN       NaN        NaN        NaN        NaN        NaN   
1   3   3  -24     30.0       NaN        NaN        NaN        NaN        NaN   
2   4  -4   60   -180.0     140.0        NaN        NaN        NaN        NaN   
3   5   5 -120    630.0   -1120.0      630.0        NaN        NaN        NaN   
4   6  -6  210  -1680.0    5040.0    -6300.0     2772.0        NaN        NaN   
5   7   7 -336   3780.0  -16800.0    34650.0   -33264.0    12012.0        NaN   
6   8  -8  504  -7560.0   46200.0  -138600.0   216216.0  -168168.0    51480.0   
7   9   9 -720  13860.0 -110880.0   450450.0 -1009008.0  1261260.0  -823680.0   
8  10 -10  990 -23760.0  240240.0 -1261260.0  3783780.0 -6726720.0  7001280.0   

          x9       x10  
0        NaN       NaN  
1        NaN       NaN  
2        NaN       NaN  
3        NaN       NaN  
4        NaN       NaN  
5        NaN       NaN  
6        NaN       NaN  
7   218790.0       NaN  
8 -3938220.0  923780.0

从数据来看，使用Cholesky分解算法计算得到的数据精度比使用Gauss消元法的精度要高，这种差异当n越大时越明显，比如当n=10时，Cholesky分解得到的x1=-9.997668，而Gauss消元法却是x1=-9.997088，误差就比Cholesky分解得到的要大。从算法上来讲，Cholesky分解及其稳定，所以Cholesky分解精度高也在意料之中。